### PR TITLE
Unit tests and github workflow

### PR DESCRIPTION
**Pull request by bencekov**
Apologies for multiple changes in one pr. This pr includes the following:
- Finished (for now) unit tests
- github workflow separated into multiple files
-  added unit tests to build workflow

### CODE REVIEW

File: main_test.go [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+// --------------------------------------------
+// TESTING WITH CORRECT SERVERS
+// --------------------------------------------
+func TestHandleCreateFlowWithoutCookie(t *testing.T) {
```
Comment (nsklikas) [inline main_test.go:21]: `testing.T` provides `Cleanup` function (https://pkg.go.dev/testing#T.Cleanup). You could pass it to `testServers.CreateTestServers` and the servers can then register their own clean up methods so that you don't have to do it in the test.

File: main_test.go [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+	if res.StatusCode != 200 {
+		t.Errorf("expected StatusCode to be 200 got %v", res.StatusCode)
+	}
+	requestLoginResponse := new(testLoginBrowser.OAuth2RequestLoginResponse)
+	if err := json.Unmarshal(data, requestLoginResponse); err != nil {
+		t.Errorf("expected error to be nil got %v", err)
+	}
+	if requestLoginResponse.Redirect_to != "test.test" {
+		t.Errorf("expected test.test, got %v", string(data))
+	}
```
Comment (nsklikas) [inline main_test.go:81]: You should use [assertions](https://pkg.go.dev/github.com/stretchr/testify/assert) to verify the response's output

File: main_test.go [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+	if err != nil {
+		t.Errorf("expected error to be nil got %v", err)
+	}
+	if res.StatusCode != 200 {
+		t.Errorf("expected StatusCode to be 200 got %v", res.StatusCode)
+	}
+	requestLoginResponse := new(testLoginBrowser.OAuth2RequestLoginResponse)
+	if err := json.Unmarshal(data, requestLoginResponse); err != nil {
+		t.Errorf("expected error to be nil got %v", err)
+	}
```
Comment (nsklikas) [inline main_test.go:78]: This is repeated in too many places, perhaps we could move this logic to a `UnamarshalJsonBody` (or something like that) function.

File: main_test.go [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+// currently only prints out results main.go needs pr to handle timeouts
+// --------------------------------------------
+func TestHandleCreateFlowTimeout(t *testing.T) {
+	data, err := CreateGenericTest(testServers.CreateTimeoutServers, http.MethodPut,
+		"/api/kratos/self-service/login/browser?aal=aal1&login_challenge=&refresh=false&return_to=http://test.test",
+		nil, handleCreateFlow)
+	if err != nil {
+		t.Errorf("expected error to be nil got %v", err)
+	}
+	t.Logf("Result:\n%s\n", string(data))
```
Comment (nsklikas) [inline main_test.go:185]: We are not asserting something in those tests. We don't gracefully handle server errors in the server yet, but if the Kratos/Hydra request times out we should just return a `500` HTTP response with some explanatory error message (for now).

Let's leave a comment about it (like `TODO: Validate response when server error handling is implemented) here.

File: main_test.go [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+		t.Errorf("expected error to be nil got %v", err)
+	}
+	bodyReader := bytes.NewReader(bodyJson)
+	data, err := CreateGenericTest(testServers.CreateTimeoutServers, http.MethodPost,
+		"/api/kratos/self-service/login?flow=1111",
+		bodyReader, handleUpdateFlow)
+	if err != nil {
+		t.Errorf("expected error to be nil got %v", err)
+	}
+	t.Logf("Result:\n%s\n", string(data))
```
Comment (nsklikas) [inline main_test.go:204]: Same

File: main_test.go [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+}
+func TestHandleKratosErrorTimeout(t *testing.T) {
+	data, err := CreateGenericTest(testServers.CreateTimeoutServers, http.MethodGet,
+		"/api/kratos/self-service/errors?id=1111",
+		nil, handleKratosError)
+	if err != nil {
+		t.Errorf("expected error to be nil got %v", err)
+	}
+	t.Logf("Result:\n%s\n", string(data))
```
Comment (nsklikas) [inline main_test.go:213]: Same

File: main_test.go [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+}
+func TestHandleConsentTimeout(t *testing.T) {
+	data, err := CreateGenericTest(testServers.CreateTimeoutServers, http.MethodGet,
+		"/api/consent?consent_challenge=test_challange",
+		nil, handleConsent)
+	if err != nil {
+		t.Errorf("expected error to be nil got %v", err)
+	}
+	t.Logf("Result:\n%s\n", string(data))
```
Comment (nsklikas) [inline main_test.go:222]: Same

File: main_test.go [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+// currently only prints out results main.go needs pr to handle errors
+// --------------------------------------------
+func TestHandleCreateFlowError(t *testing.T) {
+	data, err := CreateGenericTest(testServers.CreateErrorServers, http.MethodPut,
+		"/api/kratos/self-service/login/browser?aal=aal1&login_challenge=&refresh=false&return_to=http://test.test",
+		nil, handleCreateFlow)
+	if err != nil {
+		t.Errorf("expected error to be nil got %v", err)
+	}
+	t.Logf("Result:\n%s\n", string(data))
```
Comment (nsklikas) [inline main_test.go:236]: Same

File: main_test.go [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+		t.Errorf("expected error to be nil got %v", err)
+	}
+	bodyReader := bytes.NewReader(bodyJson)
+	data, err := CreateGenericTest(testServers.CreateErrorServers, http.MethodPost,
+		"/api/kratos/self-service/login?flow=1111",
+		bodyReader, handleUpdateFlow)
+	if err != nil {
+		t.Errorf("expected error to be nil got %v", err)
+	}
+	t.Logf("Result:\n%s\n", string(data))
```
Comment (nsklikas) [inline main_test.go:255]: Same

File: main_test.go [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+}
+func TestHandleKratosErrorError(t *testing.T) {
+	data, err := CreateGenericTest(testServers.CreateErrorServers, http.MethodGet,
+		"/api/kratos/self-service/errors?id=1111",
+		nil, handleKratosError)
+	if err != nil {
+		t.Errorf("expected error to be nil got %v", err)
+	}
+	t.Logf("Result:\n%s\n", string(data))
```
Comment (nsklikas) [inline main_test.go:264]: Same

File: main_test.go [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+}
+func TestHandleConsentError(t *testing.T) {
+	data, err := CreateGenericTest(testServers.CreateErrorServers, http.MethodGet,
+		"/api/consent?consent_challenge=test_challange",
+		nil, handleConsent)
+	if err != nil {
+		t.Errorf("expected error to be nil got %v", err)
+	}
+	t.Logf("Result:\n%s\n", string(data))
```
Comment (nsklikas) [inline main_test.go:273]: Same

File: ory_mocking/Login/login.go [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+type LoginUpdateResponse struct {
+	Session       testSession.SessionResponse `json:"session"`
+	Session_token string                      `json:"session_token"`
+}
+
+func SelfServiceLoginHandler(w http.ResponseWriter, r *http.Request) {
+	traits := testSession.IdentityTraits{Name: "TestName"}
+	identity := testSession.SessionIdentity{
+		Id:        "test1234",
```
Comment (nsklikas) [inline ory_mocking/Login/login.go:24]: Let's not hard code this in the test, other functions may need this value so you would have to hard code the same string to two places which may cause unexpected errors in the future. I think it's best to put it in a global variable, eg:
```go
SESSION_IDENTITY_ID = "test1234"
```

File: ory_mocking/Login/login.go [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+		SchemaId:  "test_identity.schema.json",
+		SchemaUrl: schemaServerURL,
+		Traits:    traits,
+	}
+	session := testSession.SessionResponse{
+		Id:       "test-1111",
```
Comment (nsklikas) [inline ory_mocking/Login/login.go:30]: Same

File: ory_mocking/Consent/consent.go [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+import (
+	"encoding/json"
+	"log"
+	"net/http"
+)
+
+type OAuth2ConsentResponse struct {
+	Challenge string `json:"challenge"`
+}
```
Comment (nsklikas) [inline ory_mocking/Consent/consent.go:11]: Why do we need to define these structs, can't we use the ones in the hydra/kratos sdk?

`hydra_client.OAuth2ConsentRequest` seems to be the same struct. The same is true for the rest of the response structs.

File: ory_mocking/Testservers/testservers.go [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+
+func createKratosMockServer() *httptest.Server {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/self-service/errors", testErrors.SelfServiceErrorsHandler)
+	mux.HandleFunc("/sessions/whoami", testSession.SessionWhoAmIHandler)
+	mux.HandleFunc("/self-service/login/browser", testLoginBrowser.SelfServiceLoginBrowserHandler)
+	mux.HandleFunc("/self-service/login", testLoginUpdate.SelfServiceLoginHandler)
+
+	s := httptest.NewServer(mux)
+	os.Setenv("KRATOS_PUBLIC_URL", s.URL+"/")
```
Comment (nsklikas) [inline ory_mocking/Testservers/testservers.go:31]:
```suggestion
	os.Setenv("KRATOS_PUBLIC_URL", s.URL)
```

File: main.go [hunk 16]
```diff
@@ -114,7 +114,7 @@ func handleCreateFlow(w http.ResponseWriter, r *http.Request) {
 		session, session_resp, e := kratos.FrontendApi.ToSession(context.Background()).
 			Cookie(cookiesToString(r.Cookies())).
 			Execute()
-
+		//added now
```
[the remark this answers is not in the document]
Comment (bencekov) [reply main.go:117]: ok

File: main_test.go (hunk 1 above)
Comment (bencekov) [reply main_test.go:21]: I did it for most tests. The error tests are "generated" with a function, which I would rather not pass the t test object to for now.

File: main_test.go (hunk 2 above)
Comment (bencekov) [reply main_test.go:81]: Changed if's to assertions.

File: main_test.go (hunk 3 above)
Comment (bencekov) [reply main_test.go:78]: I rather not. It's true that it's repeated a lot, however:
- If it's ok with you I rather stick to unmarshaling to the type instead of to some map[string]interface{} construct. I think trying with some reflection magic to send in a type for instantiating would make the tests more complicated than they need to be.
- also you don't save much lines at all. Most of these highlighted lines are actually error handling which you would still handle in the tests, because convention (and my preference is) is that functions don't handle errors just send them back to the caller.

File: main_test.go (hunk 4 above)
Comment (bencekov) [reply main_test.go:185]: If it's ok with you, I'd leave the tests as they are. I left comments in main.go, and I'll implement them at some other point.

File: ory_mocking/Login/login.go (hunk 12 above)
Comment (bencekov) [reply ory_mocking/Login/login.go:24]: I refactored most vars.

File: ory_mocking/Consent/consent.go (hunk 14 above)
Comment (bencekov) [reply ory_mocking/Consent/consent.go:11]: I replaced most of the inhouse structs. I kept the ones relating to error messages. The ory libraries store error messages as map[string]interface{} structs, and I already wrote the more comfortable structs. I would use the openapi error constructs in the main.go server.

File: ory_mocking/Testservers/testservers.go (hunk 15 above)
Comment (bencekov) [reply ory_mocking/Testservers/testservers.go:31]: done

File: ory_mocking/Testservers/testservers.go [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+
+func createSchemaMockServer() *httptest.Server {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/testschema", SchemaHandler)
+	s := httptest.NewServer(mux)
+	schema_server_url = s.URL
+	handlers.SetSchemaServerURL(s.URL)
+	handlers.SetSchemaServerURL(s.URL)
+	return s
+}
```
Comment (nsklikas) [inline ory_mocking/Testservers/testservers.go:47]: Why is this needed?

File: test_identity.schema.json [hunk 18]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (nsklikas) [inline test_identity.schema.json:1]: This file should be placed in the `ory_mocking` folder.

File: main_test.go (hunk 1 above)
Comment (nsklikas) [reply main_test.go:34]: You should pass `t` as a parameter to the test server initialization function, this will allow the server to take care of its own cleanup:
```go 
...
func TestHandleCreateFlowWithoutCookie(t *testing.T) {
        testServers.CreateTestServers(t)
        ...
}
...

func CreateTestServers() func(t *testing.T) {
	kratos = createKratosMockServer(t)
	hydra = createHydraMockServer(t)
}
...

func createKratosMockServer(t *testing.T) *httptest.Server {
       ...
       t.Cleanup(s.Close())
       return s
}
```

File: ory_mocking/Handlers/handlers.go [hunk 19]
```diff
@@ -0,0 +1,219 @@
...
+
+	hydra_client "github.com/ory/hydra-client-go/v2"
+	kratos_client "github.com/ory/kratos-client-go"
+)
+
+const (
+	BROWSER_LOGIN_ID        = "test_id"
+	BROWSER_LOGIN_TYPE      = "browser"
+	OAUTH2_SUBJECT          = "test1234"
+	REDIRECT                = "test.test"
```
Comment (nsklikas) [inline ory_mocking/Handlers/handlers.go:19]: Correct me if I'm wrong, but this is 2 variables `AUTHORIZATION_REDIRECT_URL` and `CONSENT_REDIRECT_URL`.

**Review by nsklikas — CHANGES_REQUESTED** [review]

File: ory_mocking/Testservers/testservers.go (hunk 17 above)
Comment (bencekov) [reply ory_mocking/Testservers/testservers.go:47]: The client sometimes receives an object that includes an object, a schema ID, and a schema URL. The client supposedly (according to documentation) validates the object structure against the schema URL. During debugging when some of the error messages weren't clear I implemented this schema server, and passed it's url into the http responses from the mock server.

File: main_test.go (hunk 1 above)
Comment (bencekov) [reply main_test.go:34]: done. at first I was more in favour of restricting the testing objects to the test_main module, but your solution is valid too. I changed my fork to conform to this.

File: main_test.go [hunk 20]
```diff
@@ -0,0 +1,287 @@
...
+	}
+	responseRedirect := hydra_client.NewOAuth2RedirectToWithDefaults()
+	if err := json.Unmarshal(data, responseRedirect); err != nil {
+		t.Errorf("expected error to be nil got %v", err)
+	}
+
+	if responseRedirect.RedirectTo != "test.test" {
+		t.Errorf("expected test.test, got %v", string(data))
+	}
+	assert.Equalf(t, "test.test", responseRedirect.RedirectTo, "Expected %s, got %s.", "test.test", responseRedirect.RedirectTo)
```
[the remark this answers is not in the document]
Comment (bencekov) [reply main_test.go:173]: it is. I kinda had to flip flop between refactoring the functions because of cross dependencies in the modules (good call for simplifying module layout btw.) This is a mistake. I fixed it.

File: ory_mocking/Handlers/handlers.go [hunk 21]
```diff
@@ -0,0 +1,219 @@
...
+	TRAITS_NAME             = "TestName"
+	IDENTITY_ID             = "test1234"
+	SCHEMA_ID               = "test_identity.schema.json"
+	SESSION_ID              = "test-1111"
+	ERROR_CODE              = 400
+	ERROR_STATUS            = "Bad Request"
+	ERROR_REASON            = "This is a test"
+	ERROR_MESSAGE           = "This is a test"
+	DEFAULT_ERROR_CODE      = 599
+	TEST_CHALLANGE          = "test_challange"
```
[the remark this answers is not in the document]
Comment (bencekov) [reply ory_mocking/Handlers/handlers.go:30]: I agree that test_challange is not very descriptive. I'll change it, and please let me know if you find the other constant names not sufficiently descriptive.

File: ory_mocking/Handlers/handlers.go (hunk 19 above)
Comment (bencekov) [reply ory_mocking/Handlers/handlers.go:19]: fixed

File: ory_mocking/Testservers/testservers.go (hunk 17 above)
Comment (bencekov) [reply ory_mocking/Testservers/testservers.go:47]: I checked and the client doesn't actually validate with a schema received from the url. I'll take the schemaserver out of the testing process, but I'd like to keep the function, and constructs in the code for future reference. According to the doc, the client should verify, that whatever traits object it receives conform to a schema in the schema_id and schema_url parameters.

File: test_identity.schema.json (hunk 18 above)
Comment (bencekov) [reply test_identity.schema.json:1]: done

File: ory_mocking/Testservers/testservers.go (hunk 17 above)
Comment (nsklikas) [reply ory_mocking/Testservers/testservers.go:48]: I would think that the schema would be served from Kratos

**Review by nsklikas — APPROVED** [review]

File: .github/workflows/unittest.yaml [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+
+on:
+  workflow_call:
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - name: Clean Install of Node version 18
```
Comment (gruyaume) [inline .github/workflows/unittest.yaml:11]: Can't we leverage any of those workflows:
- https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-nodejs
- https://github.com/marketplace/actions/set-up-node-using-nvm

File: .github/workflows/scan.yaml [hunk 23]
```diff
@@ -0,0 +1,20 @@
+name: Scan
```
[the remark this answers is not in the document]
Comment (massimilianogo) [reply .github/workflows/scan.yaml:1]: FYI I shared your [Nginx PR](https://github.com/canonical/nginx-rock/pull/4) with the team

[the remark this answers is not in the document]
Comment (ghislainbourgeois) [reply .github/workflows/scan.yaml:1]: This is exactly what I am proposing in the spec. At some point I expect this workflow to be centralized somewhere, but this is great to start implementing it now and see if we can add improvements!

[the remark this answers is not in the document]
Comment (bencekov) [reply .github/workflows/scan.yaml:1]: sounds good

**Review by gruyaume — APPROVED** [review]

File: .github/workflows/build.yaml [hunk 24]
```diff
@@ -7,6 +7,8 @@ on:
       - main
```
[the remark this answers is not in the document]
Comment (bencekov) [reply .github/workflows/build.yaml]: done
